xarray/core/indexes.py

```diff
@@ -61,6 +65,78 @@ def intersection(self, other):  # pragma: no cover
         raise NotImplementedError()
 
 
+def _sanitize_slice_element(x):
+    from .dataarray import DataArray
+    from .variable import Variable
+
+    if isinstance(x, (Variable, DataArray)):
+        x = x.values
```

Comment:
This will compute lazy values. Is that what we want here? Or should we do like this?
```suggestion
        x = x.data
```

Reply:
Not sure what is the use case for this, but I don't think many Xarray users have already tried providing to `.sel()` slice objects with elements that consist of large Variable or DataArray objects?

We could still check their dimensions (ndim == 0) without coercing it as Numpy arrays if necessary.

Reply:
This only comes up if somebody uses a slice object, e.g., `ds.sel(x=slice(i, j))`. In this case `i` and `j` would be loaded into memory.

I think this is _probably_ OK (it is the existing behavior) but on the other hand it would also probably be slightly safer to use `.data`, which would not implicitly convert a dask array.

Reply:
Just using `.data` wouldn't work with the code below, which expects a numpy array. Is there an easy way to get the scalar value from a dask or any duck array? I don't think so? https://github.com/dask/dask/issues/2961. But I guess we could check `len(np.shape(x)) != 0` and raise before coercing to a numpy array?

Reply:
> But I guess we could check len(np.shape(x)) != 0 and raise before coercing to a numpy array?

Hmm checking this for any `x` wouldn't work with multi-index and slice objects with tuple elements:

```python
idx = pd.MultiIndex.from_tuples([(0, 'a'), (1, 'b')]
da = xr.DataArray(idx)
da.sel(dim_0=slice((0, 'a'), (1, 'b')))

np.shape((0, 'a'))  # returns (2,)
```